### DISCUSSION TITLE
How to have a parent AbstractConcreteBase model?

### DISCUSSION THREAD

**Discussion by Maybe14** [sqlalchemy/sqlalchemy#11920] [Usage Questions] [answered]
How do I have a one-to-many relationship between a AbstractConcreteBase class with multiple polymorphic models as parent models to a child model with 1 foreign key? I know this could be done with multiple foreign keys pointed at each individual table, but is there a way to do it with just 1?

I want to do something like this:
```
class Base(DeclarativeBase):
    id: Mapped[UUID] = mapped_column(postgresql.UUID(as_uuid=True), primary_key=True)
```

```
class ParentBase(AbstractConcreteBase, Base):
    @declared_attr
    def children(cls) -> Mapped["Child"]:
        return relationship("Child", back_populates="parent")
        
class ParentMixin:
    attribute1: Mapped[str]
    attribute2: Mapped[str]
    
class ParentType1(ParentMixin, ParentBase):
    __tablename__ = "parent_type_1"
    type = "type1"
   
    __mapper_args__ = {
         "polymorphic_identity": "type1",
         "concrete": True,
     }
    
class ParentType2(ParentMixin, ParentBase):
    __tablename__ = "parent_type_2"
    type = "type2"
   
    __mapper_args__ = {
         "polymorphic_identity": "type2",
         "concrete": True,
     }
```

```
class Child(Base):
    __tablename__ = "child"
    child_attribute_1: Mapped[str]
    child_attribute_2: Mapped[str]
    
    parent_id: Mapped[UUID] = mapped_column(ForeignKey("parent_base.id"))
    parent: Mapped["ParentBase"] = relationship(back_populates="children")
```

When running an alembic migration, I get the error:
```
sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'child.parent_id' could not find table 'parent_base' with which to generate a foreign key to target column 'id'
```

Which is because `ParentBase` has no `__tablename__`. And including a table name to this model will have it create `ParentBase` table which I don't want.

So then I changed `ForeignKey("parent_base.id")` to `ForeignKey(ParentBase.id)`, but that lead to this new error:
```
sqlalchemy.exc.InvalidRequestError: Can't copy ForeignKey object which refers to non-table bound Column Column('id', Uuid(), table=None, primary_key=True, nullable=False)
```

Basically instead of having to do something like this:
```
class Child(Model):
    __tablename__ = 'child'
    child_attribute_1: Mapped[str]
    child_attribute_2: Mapped[str]
    
    parent_type_1_id: Mapped[Optional[UUID]] = mapped_column(ForeignKey("parent_type_1.id"))
    parent_type_2_id: Mapped[Optional[UUID]] = mapped_column(ForeignKey("parent_type_2.id"))

    @hybrid_property
    def parent_id(self):
        return self.parent_type_1_id or self.parent_type_2_id
```

I want to combine the parent_ids into one `parent_id` foreign key

Is what I'm looking for possible?

**Comment by zzzeek** [comment]
hi -

no, the `ForeignKey` represents semantically the identical thing as a real `FOREIGN KEY` constraint in a relational database.   it points to a column in a table.   you can't use the name of a table that does not exist.

when you configure columns and foreign keys on the ORM, always keep those literal as to what they really are in the database.  in this case you want to make a separate relationships on `ParentType1` and `ParentType2` that each link to `Child`, using separate columns on the 'child' table.   if you then want to load a `Child` and get the "parent" object, **then** you can have a plain python `@property` called `parent` that does `return self.parent_relationship_one or self.parent_relationship_two`, or whatever conditional you want it to use in order to decide.

**Comment by Maybe14** [comment, marked as answer]
I was able to work around this by using `primaryjoin` and not using an explicit `ForeignKey`

```
class ParentBase(AbstractConcreteBase, Base):
    @declared_attr
    def children(cls) -> Mapped["Child"]:
        return relationship(
            "Child", 
            back_populates="parent", 
            primaryjoin=lambda: and_(foreign(Child.parent_id) == cls.id)  # not in quotes to be able to use cls
        )
```

```
class Child(Base):
    __tablename__ = "child"
    child_attribute_1: Mapped[str]
    child_attribute_2: Mapped[str]
    
    parent_id: Mapped[UUID]
    parent: Mapped["ParentBase"] = relationship(
        back_populates="children", 
        primaryjoin="and_(foreign(Child.parent_id)==Parent.id)"  # in quotes to avoid circular import
    )
```

With this I'm able to do `parent.children` and `child.parent`
